contracts/exchange/api/AssertionContract.sol

```diff
@@ -0,0 +1,404 @@
+        "OnboardedTransferAccount mismatch"
+      );
+    }
+
+    for (uint i = 0; i < assertion.signers.length; i++) {
+      require(
+        account.signers[assertion.signers[i].signer] == assertion.signers[i].permission,
+        "Signer permission mismatch"
+      );
+    }
+  }
+
+  function assertSubAccount(SubAccountAssertion calldata assertion, bool requireFullPositionMap) public view {
+    SubAccount storage subAccount = state.subAccounts[assertion.id];
+
+    require(assertion.id != 0, "id == 0 in SubAccount assertion");
+    require(subAccount.id != 0, "SubAccount not found");
+    require(subAccount.id == assertion.id, "Unexpected SubAccount ID");
+    require(subAccount.adminCount == assertion.adminCount, "AdminCount mismatch");
+    require(subAccount.signerCount == assertion.signerCount, "SignerCount mismatch");
+    require(subAccount.accountID == assertion.accountID, "AccountID mismatch");
+    require(subAccount.marginType == assertion.marginType, "MarginType mismatch");
+    require(subAccount.quoteCurrency == assertion.quoteCurrency, "QuoteCurrency mismatch");
+    require(
+      subAccount.lastAppliedFundingTimestamp == assertion.lastAppliedFundingTimestamp,
+      "LastAppliedFundingTimestamp mismatch"
+    );
+
+    for (uint i = 0; i < assertion.spotBalances.length; i++) {
+      require(
+        subAccount.spotBalances[assertion.spotBalances[i].currency] == assertion.spotBalances[i].balance,
+        "SpotBalance mismatch"
+      );
+    }
+
+    assertPositionsMap(subAccount.options, assertion.options, "Options", requireFullPositionMap);
+    assertPositionsMap(subAccount.futures, assertion.futures, "Futures", requireFullPositionMap);
+    assertPositionsMap(subAccount.perps, assertion.perps, "Perps", requireFullPositionMap);
+
+    for (uint i = 0; i < assertion.signers.length; i++) {
+      require(
+        subAccount.signers[assertion.signers[i].signer] == assertion.signers[i].permission,
+        "Signer permission mismatch"
+      );
+    }
+  }
+
+  function assertPositionsMap(
+    PositionsMap storage posMap,
+    Position[] memory assertion,
+    string memory mapName,
+    bool requireFullPositionMap
+  ) internal view {
+    if (requireFullPositionMap) {
+      require(
+        posMap.keys.length == assertion.length,
+        string(abi.encodePacked(mapName, " PositionsMap length mismatch"))
+      );
+    }
+
+    for (uint i = 0; i < assertion.length; i++) {
+      Position storage pos = posMap.values[assertion[i].id];
+      require(pos.balance == assertion[i].balance, string(abi.encodePacked(mapName, " Position balance mismatch")));
+      require(
+        pos.lastAppliedFundingIndex == assertion[i].lastAppliedFundingIndex,
+        string(abi.encodePacked(mapName, " Position lastAppliedFundingIndex mismatch"))
+      );
+    }
+  }
+
+  function isAllAccountExists(address[] calldata accountIDs) public view returns (bool) {
+    for (uint256 i = 0; i < accountIDs.length; i++) {
```

Comment:
limit the number of accountIDs passed in to 500 to prevent the quadratic for loop issue